pennylane/measurements/probs.py

```diff
@@ -259,23 +259,28 @@ def _count_samples(indices, batch_size, dim):
         """Count the occurrences of sampled indices and convert them to relative
         counts in order to estimate their occurrence probability."""
         num_bins, bin_size = indices.shape[-2:]
-        if batch_size is None:
-            prob = qml.math.zeros((dim, num_bins), dtype="float64")
-            # count the basis state occurrences, and construct the probability vector for each bin
+        interface = qml.math.get_deep_interface(indices)
+
+        def _count_samples_core(indices, dim, interface):
+            if qml.math.is_abstract(indices):
+                return qml.math.array(
+                    [[qml.math.sum(idx == p) for idx in indices] for p in range(dim)],
+                    like=interface,
+                )
+            probabilities = qml.math.zeros((dim, num_bins), dtype="float64")
             for b, idx in enumerate(indices):
                 basis_states, counts = qml.math.unique(idx, return_counts=True)
-                prob[basis_states, b] = counts / bin_size
-
-            return prob
+                probabilities[basis_states, b] = counts
+            return probabilities
```

Comment:
As we know `indices` at the time we define this function, we could also define the function differently, conditioned on whether `qml.is_abstract(indices)`, rather than having one function that decides between the two cases.
This is a rather subjective suggestion, so feel free not to go with it!

Reply:
Do you mean something like
```
        if qml.math.is_abstract(indices):
            def _count_samples_core(indices, dim, interface):
                return qml.math.array(
                    [[qml.math.sum(idx == p) for idx in indices] for p in range(dim)],
                    like=interface,
                )
        else:
            def _count_samples_core(indices, dim, interface):
                probabilities = qml.math.zeros((dim, num_bins), dtype="float64")
                for b, idx in enumerate(indices):
                    basis_states, counts = qml.math.unique(idx, return_counts=True)
                    probabilities[basis_states, b] = counts
                return probabilities
```
?

Reply:
Yep, exactly :)